docs/madr/decisions/046-grpc-tcp-probes.md

```diff
@@ -0,0 +1,289 @@
+
+Handling incoming probe request from kubelet:
+
+1. If the user defines an HTTP probe, requests from kubelet are forwarded/redirected to application directly with necessary path/port rewriting, etc.
+
+2. If the user defines a gRPC probe, we'll also receive HTTP requests from kubelet, since they are transformed to HTTP probes. We forward these requests to the translator and the translator is responsible for handling the probe request, by detecting the actual status from the application.
+
+The whole workflow will be like the following diagram:
+
+```
+  kubelet (TCP Probes)                                   HTTP to TCP translator  --->  Application (TCP port)
+               ↘                                                ↗ 
+                 Pod (HTTP Probes)  --->  Virtual Probe Listener (HTTP routing)
+               ↗                                                ↘
+  kubelet (gRPC Probes)                                  HTTP to gRPC translator  --->  Application (gRPC healthcheck)
+```
+
+#### Positive Consequences
+
+- Don't need to introduce new ports for gRPC probes
+
+- Better user experience: don't need to introduce new annotations, the user can still specify the port as they are doing now.
+
+- Less virtual listeners on sidecar and saves a port to be taken from application: the translator will be used to handle both TCP probes and gRPC probes
+
+- Don't need to introduce new fields on the `Dataplane` object
+
+#### Negative Consequences
+
+- Requires an extra component (the translator) in `kuma-dp`, which increases the implementation complexity
+
+- Less intuitiveness, harder for troubleshooting issues
+
+#### Implementation details
+
+The probes support can be implemented in the following steps:
+
+1. Change the `Dataplane` model to contain fields for gRPC and TCP probes (see examples below)
+2. Create a new component in `kuma-dp` (the probe translator) to handle the incoming HTTP probes by performing the actual gRPC and TCP probes and returning the result
+3. Generate a new Envoy cluster `kuma-virtual-probes` and point it to the component
+4. Make sidecar injector to:
+   1. Get the correct port for the probe translator, and generate corresponding command line arguments for `kuma-dp` 
+   2. Identify gRPC and TCP probes and transform them into the updated `Dataplane` object
+5. Change the route of Virtual Probe listener (in `probe_generator`) and forward the user defined gRPC and TCP probes into the new component in `kuma-dp`
+
+In step 2, the new component is an HTTP server and needs to listen on a new port, a different one from the existing Virtual Probe listener port (default `9000`). The new port will be configurable by `runtime.kubernetes.injector.virtualProbesPort` and the default value is `9001`. The user is not expected to change this port unless it is conflicting with an application port. 
```

Comment:
why do we need to change the port?

Reply:
I made a mistake on the configuration key, it should be a new and different key.
We need a new port because we'll need to listen to it in `kuma-dp`, and the current one (default `9000`) was taken by Envoy, so we can't listen on that same port.

Reply:
technically we can. For example with Envoy Admin we listen on localhost, but we expose parts of this on PodIP network interface. We could do the same here. However, this problem is gone if we consider skipping port with iptables https://github.com/kumahq/kuma/pull/9964#discussion_r1591939749

Reply:
Yeah, let me add it into the doc.

Reply:
I'll mark this conversation as resolved since it's now in the proposed approach.